plugins/languages/typescript-tools.nix

```diff
@@ -0,0 +1,177 @@
+{
+  lib,
+  pkgs,
+  helpers,
+  config,
+  ...
+}:
+with lib; let
+  cfg = config.plugins.typescript-tools;
+in {
+  options.plugins.typescript-tools = {
+    enable = mkEnableOption "typescript-tools";
+    package =
+      helpers.mkPackageOption "typescript-tools"
+      pkgs.vimPlugins.typescript-tools-nvim;
+
+    handlers = mkOption {
+      type = with helpers.nixvimTypes; nullOr (attrsOf strLuaFn);
+      apply = v: helpers.ifNonNull' v (mapAttrs (_: helpers.mkRaw) v);
+      default = null;
+      description = "How tsserver should respond to LSP requests";
+      example = {
+        "textDocument/publishDiagnostics" = ''
+          require("typescript-tools.api").filter_diagnostics(
+            -- Ignore 'This may be converted to an async function' diagnostics.
+            { 80006 }
+          )
+        '';
+      };
+    };
+
+    settings = {
+      separateDiagnosticServer =
+        helpers.defaultNullOpts.mkBool true
+        "Spawns an additional tsserver instance to calculate diagnostics";
+
+      publishDiagnosticOn = mkOption {
+        type = with types; enum ["change" "insert_leave"];
+        default = "insert_leave";
+        description = ''
+          Either "change" or "insert_leave". Determines when the client asks the server about diagnostics'';
+      };
+
+      exposeAsCodeAction = mkOption {
+        type = with types;
+          either (enum ["all"]) (enum [
```

Comment:
you should be able to merge the two enum, though it should be a `helpers.defaultNullOpts.mkListOf (enum ....)`

Reply:
This plugin's docs state that it has to be either "all" or any number of the other options. If I merge these, wouldn't that allow something incorrect like `["all", "fix_all"]`?

Reply:
Yes it would. You need something of the form (either (enum ["all"]) (listOf (enum [....])))